src/Plutarch/Extra/Bind.hs

```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Extra.Bind (
+    -- * Type class
+    PBind (..),
+
+    -- * Functions
+    (#>>=),
+    pjoin,
+) where
+
+import Data.Kind (Type)
+import Plutarch (
+    S,
+    Term,
+    pcon,
+    pfix,
+    phoistAcyclic,
+    plam,
+    unTermCont,
+    (#),
+    (#$),
+    type (:-->),
+ )
+import Plutarch.Api.V1.Maybe (PMaybeData (PDJust, PDNothing))
+import Plutarch.Builtin (PBuiltinList, pfromData)
+import Plutarch.DataRepr (pfield)
+import Plutarch.Either (PEither (PLeft, PRight))
+import Plutarch.Extra.Applicative (PApply)
+import Plutarch.Extra.Function (pidentity)
+import Plutarch.Extra.Functor (PSubcategory)
+import Plutarch.Extra.TermCont (pmatchC)
+import Plutarch.List (PList, pconcat, pnil, puncons)
+import Plutarch.Maybe (PMaybe (PJust, PNothing))
+import Plutarch.Pair (PPair (PPair))
+
+{- | Gives the capability to bind a Kleisli arrow over @f@ to a value:
+ essentially, the equivalent of Haskell's '>>='. Unlike Haskell, we don't
+ require the availability of 'pure': to recover the equivalent of Haskell's
+ 'Monad', you want both 'PApplicative' and 'PBind'.
+
+ = Laws
+
+ * @'pbind' '#' ('pbind' '#' m '#' f) '#' g@ @=@
+ @'pbind' '#' m '#' ('plam' '$' \x -> 'pbind' '#' (f '#' x) '#' g)@
+ * @'pbind' '#' f '#' ('plam' '$' \g -> 'pfmap' '#' g '#' x)@ @=@
+ @'pliftA2' '#' 'papply' '#' f '#' x@
+
+ @since 1.2.1
+-}
+class (PApply f) => PBind (f :: (S -> Type) -> S -> Type) where
+    -- | '>>=', but as a Plutarch function.
+    pbind ::
+        forall (a :: S -> Type) (b :: S -> Type) (s :: S).
+        (PSubcategory f a, PSubcategory f b) =>
+        Term s (f a :--> (a :--> f b) :--> f b)
```

Comment:
I think you'd want bind to be inlined most of the time (and inline its arguments into its body), so I suppose a Haskell function would be better.

Reply:
A Haskell function would actually _write_ better as well: `#>>=` versus `pbind` is not only more familiar-looking, but leads to much more readable code. To see how far this goes, just look at the laws! So I agree, definitely want to make that change.